---
editor_options: 
  markdown: 
    wrap: 72
---

# scRNAseq-Tutorial

This repo contains a prepared counts and meta data file from

Ximerakis, M., Lipnick, S.L., Innes, B.T. *et al.* Single-cell
transcriptomic profiling of the aging mouse brain. *Nat Neurosci*
**22,** 1696--1708 (2019). <https://doi.org/10.1038/s41593-019-0491-3>

GSE: GSE129788

We are looking at one of the single cell RNAseq samples.

<br />

If you want to code along, please install these packages to RStudio
before class. It could take up to 10-15 min.

`Install.packages("Seurat")`

`Install.packages("ggplot2")`

<br />

Download the data we will be working with. You need these files:

*Ximerakis\_tutorial.Rmd*

*counts\_file\_1.txt*

*metadata\_file\_1.txt*

<br />

Google drive:
<br />
<https://drive.google.com/drive/folders/1MkU_yQT6fWZmkq3c8nHRoLWXh4L2-pFm?usp=sharing>

<br />

GitHub repo:
<br />
<https://github.com/mcap91/scRNAseq-Tutorial>

<br />

Advanced:

Install and configure git bash, and enable R studio to use Git terminal
<br />
<https://www.neonscience.org/setup-git-bash-rstudio>
<br />
<https://www.stanleyulili.com/git/how-to-install-git-bash-on-windows/>

Clone the above GitHub repository into a new directory on your local
machine.

`cd` to a directory where the repository will be copied

`git clone https://github.com/mcap91/scRNAseq-Tutorial.git`

<br />

Open up the `.Rmd` file to follow along with the class.

Open up the `.Rproj` file to easily edit contents for add/commit/push
functionality

<br />
<br />
